src/environment-logs-provider.ts

```diff
@@ -0,0 +1,168 @@
+import * as vscode from "vscode";
+import { Environment } from "./env0-environments-provider";
+import { apiClient } from "./api-client";
+import stripAnsi from "strip-ansi";
+import { DeploymentStatus, DeploymentStepStatus } from "./types";
+import axios from "axios";
+
+const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
+const pollStepLogsInterval = 1000;
+
+export class EnvironmentLogsProvider {
+  private static environmentLogsOutputChannel: vscode.OutputChannel;
+  private readonly abortController = new AbortController();
+  private _isAborted = false;
+  private readonly stepsAlreadyLogged: string[] = [];
+  constructor(private readonly env: Environment) {
+    this.logDeployment(this.env.latestDeploymentLogId).catch((e) => {
+      if (!axios.isCancel(e)) {
+        throw e;
+      }
+    });
+  }
+
+  abort() {
+    this.abortController.abort();
+    this._isAborted = true;
+    EnvironmentLogsProvider.environmentLogsOutputChannel.clear();
+  }
+
+  get isAborted() {
+    return this._isAborted;
+  }
+
+  private log(value: string) {
+    if (!this.isAborted) {
+      const channel = EnvironmentLogsProvider.environmentLogsOutputChannel;
+      if (!channel) {
+        throw new Error(
+          "environment logs output channel used before initialized"
+        );
+      }
+      channel.appendLine(value);
+    }
+  }
+
+  private async logDeployment(deploymentId: string) {
+    let previousStatus;
+    EnvironmentLogsProvider.environmentLogsOutputChannel.show();
+
+    const deployment = await apiClient.getDeployment(
+      deploymentId,
+      this.abortController
+    );
+
+    if (deployment.status === DeploymentStatus.QUEUED) {
+      this.log(`Deployment is queued! Waiting for it to start...`);
+    }
+
+    while (!this.isAborted) {
```

Comment:
extract `while` content, and use `setInterval` instead of sleeping for delaying next execution

Reply:
I have concerns about using `setInterval`. In the current implementation with the `while` loop, I can use await to ensure that processDeploymentSteps completes before proceeding. This offers controlled management. However, with setInterval, there's uncertainty regarding how long processDeploymentSteps might take to finish. This could make synchronization and management more challenging, especially since other internal functions like writeDeploymentStepLog also wait for certain steps to complete.

Reply:
👍